### ISSUE TITLE
Questions

### ISSUE THREAD

**Issue by the-database** [2minkyulee/AESOP-Auto-Encoded-Supervision-for-Perceptual-Image-Super-Resolution#1]
Hi, thanks for your interesting research. I am attempting to implement the AESOP loss in my SR framework, but I am seeing some rainbow artifacts in my SR outputs, so I have a few questions. 

When training the AE, what batch size is used? 

Besides RRDB, the paper shows successful results with other backbones such as SwinIR and DRCT. For these tests, does the backbone refer to the AE decoder backbone, the SR backbone, or both? Is a distinct AE trained for each backbone? Or could an RRDB-based AE be used to train a SwinIR SR model, for example?

For real-world SR, can the bicubic trained AE be used, or does an AE need to be trained specifically with the same real world degradations for the LR dataset as the real world SR model? 

While the paper evaluates AESOP loss in the context of perceptual SR, could it also be used to improve fidelity oriented SR, by training a model solely with AESOP loss? Or does AESOP only make sense in combination with a perceptual loss? 

Thanks in advance.

**Comment by 2minkyulee** [comment]
Thank you for your interest in our work. Below are details.

## 1.Training AE
- Batchsize 64
- GT size 256
- DF2K 
- MATLAB bicubic down (double precision) from BasicSR scripts
- Decoder initialized with pretrained RRDB (official weight)

We have tried other settings, (e.g. substitute encoder with torch interpolation) but did not observe rainbow artifacts.

## 2. Backbone
Backbone indicates the generator (SR network) architecture.
For AE, we fix it as RRDB for simplicity and efficiency (since Transformer-based are often slow and memory intensive).

We did not test Swin-AE or DRCT-AE, but we believe that it will work fine.

## 3. Realworld SR
RealWorld trained AE works better. 
We also initialized the Decoder with a Realworld version RRDB (RealESRNet, official weight)

## 4. AESOP on fidelity SR
The AESOP objective,  uses the fidelity bias of the GT for reconstruction.
This shares some intuition with ECO (https://arxiv.org/abs/2312.17526).

The shared intuition is that both are using a less noisy term as the target image, which may help regression.
Thus, AESOP may lead to performance gain in only-fidelity SR settings.

But these are not tested.

**Comment by the-database** [comment]
Thanks for clarifying your training methods. I was using bicubic trained AE for real world SR, so I will try again with real world AE and see if that helps. 

Thanks for clarifying the link to your previous research in ECO. In general I see much higher fidelity with AE than a simple pretrained SR network, so `AE(gt)` may be a more accurate estimate of the optimal centroid which could improve the results of the original ECO method. I have not tried implementing ECO yet, but this seems worth exploring.

To more easily debug my issue with rainbow artifacts, I have been exploring AESOP on fidelity SR since it's a simpler case than perceptual. In this case, I am running into an issue where, when training SR from scratch optimizing AESOP loss only, the artifacts commonly observed in early training don't go away after a few thousand iterations, even though they do go away in a few thousand iterations when optimizing on L1 loss only. I visualized the outputs of `predicted`, `GT`, `AE(predicted)`, `AE(GT)`, and found that `predicted` produces artifacts as expected when training SR from scratch in early iterations, but then `AE(predicted)` removes these artifacts. My intuition is that since AE removes the artifacts from the predicted output, the model never learns to correct these artifacts. 

Does my finding suggest that AESOP loss should not be used to optimize an SR model from scratch? Or does it suggest some issue with my AE model? Or maybe some other bug in my implementation. For example, predicted outputs often fall outside of the [0, 1] range, especially in early training, is it necessary to use any clamping to handle this, or are the AE model and AESOP loss well behaved without any clamping?

Here are some visual examples. Notice for the AESOP loss example, the artifacts in the SR predicted output vs the lack of artifacts in AE(SR predicted output). In the same number of iterations, for the L1 loss example, the artifacts are significantly reduced in the SR predicted output. 

**Training RRDBNet from scratch on DF2K dataset, optimizing L1 loss, after 2000 iterations**:
SR predicted output vs SR target output:

![Image](https://github.com/user-attachments/assets/b5ff342a-f47c-4fdb-83c5-3ec750d13aba)
![Image](https://github.com/user-attachments/assets/28fdcdf3-e5e4-40a4-9ad2-1d5aee07c697)

**Training RRDBNet from scratch on DF2K dataset, optimizing AESOP loss, after 2000 iterations**:
SR predicted output vs AE(SR predicted output) vs SR target output vs AE(SR target output):


![Image](https://github.com/user-attachments/assets/c3f0a6e6-9542-4773-9995-0a7ee2251909)
![Image](https://github.com/user-attachments/assets/15f71358-2035-456f-bef8-075c87e42bb9)
![Image](https://github.com/user-attachments/assets/2cb319bb-8b3e-4bef-b41e-d71122783ff8)
![Image](https://github.com/user-attachments/assets/fd8e814d-aa31-4d1e-a5bd-2f96fda3ae5f)

**Comment by 2minkyulee** [comment]
**# 0**
When training RealWorld SR nets, we normally start with bicubic pretrained SR nets.
Starting with a network that can produce sharp details is often helpful, and the performance gain is significant.
https://arxiv.org/abs/2007.07524

I would suggest starting from a bicubic sr net, instead of training from scratch.
The typical procedure is as follows.

1. Train x2 bicubic SR
2. Finetune for x4 bicubic SR (replace the last upscale head to match img size)
3. Finetune for real-world x4 SR


---
**# 1**
We haven't tested AESOP for training SR nets from scratch.
Since AE is frozen, it might be vulnerable to ood distribution pixels / img patches (e.g., pixels that fall outside [0, 1]).
- Maybe an appropriate norm before the AE might help (e.g., min-max norm, w/ identical params for both SR and GT). 
- Or slightly fine-tuning the AE to recon the noisy SR images on the fly might help. (be aware of collapsing)

But these are not tested.

---
**# 2**
Are you using perceptual loss (lpips loss) or Artifact loss (LDL loss) together? 
I believe these may help.

---
**# 3**
I expect that manually designing an AE with a lower compression ratio will also help.
For example, you could consider using x2 AE.
But this will probably lead to lower perceptual scores. 

---
**# 4**
If your project is not for a research purpose
and if you don't have to strictly follow the MATLAB bicubic downsampling scheme,
and you simply want to train fidelity SR nets,

I would suggest simply using ECO on the fly (which I expect to be more stable than AESOP scratch),
- the pseudo-codes for on-the-fly ECO are in the official repository.
- and to the best of my knowledge, neosr supports ECO (https://github.com/neosr-project/neosr)

**Comment by the-database** [comment]
Thanks for your insights, this gives me several things to test. First I'll attempt to reproduce the results in your paper for the bicubic SR task, following your experiment settings as closely as possible, which should tell me if my implementation has any errors. If I find the cause of the artifact issue I'm seeing, I'll post it here. Once AESOP loss produces the expected results for bicubic SR, I can explore real world SR or any possible improvements to fidelity oriented SR with ECO. Thanks again for your help.

**Comment by Baekhyunn** [comment]
hey, could u share AE and AESOP loss? Thanks in advance.

**Comment by 2minkyulee** [comment]
@Baekhyunn Codes for AESOP loss and the pretrained AE weights are updated!

**Comment by the-database** [comment]
Thank you for posting the code and weights!
